R/quote.R

```diff
@@ -0,0 +1,258 @@
+#'   function) and an environment (the original call site of the
+#'   function), and they self-evaluate to return a value as soon as
+#'   you touch them. Similarly, formulas self-evaluate when called
+#'   within \code{\link{f_eval}()}. However, unlike promises, they are
+#'   first-class objects: you can pass a formula around and use it to
+#'   transform another formula or expression. Formulas are thus
+#'   treated as reified promises.
+#'
+#' @section Theory: Formally, \code{f_quote()} and \code{expr_quote()}
+#'   are quasiquote functions, \code{UQ()} is the unquote operator,
+#'   and \code{UQS()} is the unquote splice operator. These terms
+#'   have a rich history in LISP, and live on in modern languages like
+#'   \href{Julia}{http://docs.julialang.org/en/release-0.1/manual/metaprogramming/} and
+#'   \href{Racket}{https://docs.racket-lang.org/reference/quasiquote.html}.
+#'
+#' @param expr An expression.
+#' @param x A function, raw expression, or formula to interpolate.
+#'   When interpolating a formula, other formulas are treated as
+#'   promises (see section on tidy evaluation).
+#' @return For \code{f_quote()}, a formula (which contains information
+#'   about where to find the objects mentioned in the formula). For
+#'   \code{expr_quote()}, a raw quoted expression.
+#' @export
+#' @aliases UQ UQE UQF UQS
+#' @examples
+#' # When a tidyeval function captures an argument, it is wrapped in a
+#' # formula and interpolated. f_quote() is a simple wrapper around
+#' # arg_capture() and as such is the fundamental tidyeval
+#' # function. It allows you to quote an expression and interpolate
+#' # unquoted parts:
+#' f_quote(foo(bar))
+#' f_quote(!! 1 + 2)
+#' f_quote(paste0(!! letters[1:2], "foo"))
+#'
+#' # Alternatively you can interpolate a formula that is already
+#' # constructed:
+#' interp(~!! 1 + 2)
+#' f <- ~paste0(!! letters[1:2], "foo")
+#' interp(f)
+#'
+#' # The !! operator is a syntactic shortcut for unquoting. However
+#' # you need to be a bit careful with operator precedence. All
+#' # arithmetic and comparison operators bind more tightly than `!`:
+#' interp(x ~ 1 +  !! (1 + 2 + 3) + 10)
+#'
+#' # For this reason you should always wrap the unquoted expression
+#' # with parentheses when operators are involved:
+#' interp(x ~ 1 + (!! 1 + 2 + 3) + 10)
+#'
+#' # Or you can use the explicit unquote function:
+#' interp(x ~ 1 + UQ(1 + 2 + 3) + 10)
+#'
+#'
+#' # Use !!! or UQS() if you want to add multiple arguments to a
+#' # function It must evaluate to a list
+#' args <- list(1:10, na.rm = TRUE)
+#' interp(~mean(!!! args))
+#' f_quote(mean( UQS(args) ))
+#'
+#' # You can combine the two
+#' var <- quote(xyz)
+#' extra_args <- list(trim = 0.9, na.rm = TRUE)
+#' f_quote(mean(UQ(var) , UQS(extra_args)))
+#' interp(~mean(!!var , !!!extra_args))
+#'
+#'
+#' # Unquoting is especially useful for transforming a captured
+#' # expression:
+#' expr <- quote(foo(bar))
+#' expr <- expr_quote(inner(!! expr, arg1))
```

Comment:
Maybe `expr_quote()` should go in another documentation file?  It's going to be used in a slightly different context.